src/input/joystick.c

```diff
@@ -0,0 +1,714 @@
+#include "joystick.h"
+
+#include "core/log.h"
+#include "core/speed.h"
+#include "core/time.h"
+#include "game/system.h"
+#include "graphics/window.h"
+#include "input/hotkey.h"
+#include "input/keyboard.h"
+#include "input/mouse.h"
+#include "input/scroll.h"
+
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+
+typedef int joystick_axis;
+typedef int joystick_button;
+
+enum {
+    DIRECTION_UP = 0,
+    DIRECTION_LEFT = 1,
+    DIRECTION_DOWN = 2,
+    DIRECTION_RIGHT = 3,
+    NUM_DIRECTIONS = 4
+};
```

Comment:
Any reason why this is not a `typedef enum { ...} joystick_button`?

Reply:
Because the button is independent from the directions. The button can be an `A` button, a `X` button, etc... Directions are used in entirely different places, mainly to calculate the axis direction.